Sources/SpeziAccount/Views/UserIdPasswordEmbeddedView.swift

```diff
@@ -0,0 +1,161 @@
+                Text("UP_LOGIN", bundle: .module)
+                    .padding(8)
+                    .frame(maxWidth: .infinity)
+            }
+                .buttonStyle(.borderedProminent)
+                .environment(\.defaultErrorDescription, .init("UP_LOGIN_FAILED_DEFAULT_ERROR", bundle: .atURL(from: .module)))
+                .padding(.bottom, 12)
+                .padding(.top)
+
+
+            HStack {
+                Text("UP_NO_ACCOUNT_YET", bundle: .module)
+                NavigationLink {
+                    service.viewStyle.makeSignupView()
+                } label: {
+                    Text("UP_SIGNUP", bundle: .module)
+                }
+            }
+                .font(.footnote)
+        }
+            .disableDismissiveActions(isProcessing: state)
+            .viewStateAlert(state: $state)
+            .sheet(isPresented: $presentingPasswordForgetSheet) {
+                NavigationStack {
+                    service.viewStyle.makePasswordResetView()
+                        .navigationBarTitleDisplayMode(.inline)
+                }
+            }
+            .onTapGesture {
+                focusedField = nil
+            }
+    }
+
+
+    @ViewBuilder private var fields: some View {
+        VStack {
+            Group {
+                VerifiableTextField(userIdConfiguration.idType.localizedStringResource, text: $userId)
+                    .environmentObject(userIdValidation)
+                    .textContentType(userIdConfiguration.textContentType)
+                    .keyboardType(userIdConfiguration.keyboardType)
+                    .onTapFocus(focusedField: _focusedField, fieldIdentifier: .userId)
+                    .padding(.bottom, 0.5)
+
+                VerifiableTextField(.init("UP_PASSWORD", bundle: .atURL(from: .module)), text: $password, type: .secure) {
+                    Button(action: {
+                        presentingPasswordForgetSheet = true
+                    }) {
+                        Text("UP_FORGOT_PASSWORD", bundle: .module)
+                            .font(.caption)
+                            .bold()
+                            .foregroundColor(Color(uiColor: .systemGray))
+                    }
+                }
+                    .environmentObject(passwordValidation)
+                    .textContentType(.password)
+                    .onTapFocus(focusedField: _focusedField, fieldIdentifier: .password)
+            }
+                .disableFieldAssistants()
+                .textFieldStyle(.roundedBorder)
+                .font(.title3)
+        }
+    }
+
+
+    /// Create a new embedded view.
+    /// - Parameter service: The ``UserIdPasswordAccountService`` instance.
+    public init(using service: Service) {
+        self.service = service
+    }
+
+
+    private func loginButtonAction() async throws {
+        userIdValidation.runValidation(input: userId)
+        passwordValidation.runValidation(input: password)
```

Comment:
To Illustrator the background on this maybe. There are multiple parts to this code section.

In the general case, all validations are run while the user types (with a small delay). However, we don't show the `nonEmpty` validation error if the user erases their text and show the error only if they try to submit their input (while navigating focus to the first field that has a failed validation). The goal here was to not annoy the user with validation messages if they are still changing their input. Consequentially, I didn't go with disabling the button in this scenarios as it would have made it impossible to discover which fields are required and which are optional and can be left empty. This is the only issue I see with the proposed change (the "previous version") where one has to input something into all fields to discover if a validation error occurs on empty input and the button is eventually enabled.

For the Login view specifically, we only check for emptiness of the input and not for any structural validations and therefore run the validations on button press only. Either the credential exists as reported by the account service later on or they don't. Makes it easier to adapt to e.g. changing password requirements and avoid issues with client-side validations.

As this is mainly about aesthetics I would propose to address this in a future PR and track this as an issue for now and maybe open this up for further discussion?

Reply:
Thank you for the context!

The behaviour that `nonEmpty` errors are only shown on submission aligns with the behaviour to only enable the button when all criteria are met. We could make the button just look disabled and when pressing it, the focus will jump to the first view that is empty and highlight the errors there. I think that would be a good combination of the typical behaviour of a button turning actionable as soon as all necessary fields are filled out, indicating to the user that they may but do not need to fill to the rest and enabling the behaviour to jump to the missing fields in case the user presses the button anyways (out of frustration or just to see the error message).
This also applies to the login views: If both of them are empty, having a grayed out login button also showcases to the user that there is some action to be performed before the button can be pressed. There is no need to include the password requirements in the password field. I think this is also fine for the username field, but that might be configurable to guide the user what type of credentials are expected.

Would the "fake" disable functionality and UI indications be a lot of work for this PR? If not, I would vote to bring it into this PR. If it would be a substantial amount of work I am fine with button it into a separate issue and then bringing it up in a separate PR 👍

Reply:
Latest changes in https://github.com/StanfordSpezi/SpeziAccount/pull/7/commits/cb714a68ff7f8653991a65420b1ba8c135a25206 incorporate the discussed feedback.